samigo/samigo-impl/src/java/org/sakaiproject/samigo/impl/SamigoAvailableNotificationServiceImpl.java

```diff
@@ -0,0 +1,369 @@
+                        if (!StringUtils.isEmpty(extension.getUser())){ //when the extension has an individual user
+                            sendEmailNotification(site, publishedAssessment, site.getMember(extension.getUser()), extension);
+                        }
+                        if (!StringUtils.isEmpty(extension.getGroup())){    //extension for a group
+                            for(Member member: site.getGroup(extension.getGroup()).getMembers()){
+                                sendEmailNotification(site, publishedAssessment, member, extension);
+                            }
+                        }
+                    }
+                } else if (StringUtils.equals(publishedAssessment.getAssessmentAccessControl().getReleaseTo(), "Selected Groups")){ //when there is a releaseTo setting that limits the access of the test
+                    for (Object groupId : publishedAssessment.getReleaseToGroups().keySet().toArray()){ //loop through applicable group IDs
+                        for (Member member : site.getGroup((String) groupId).getMembers()){  //loop through users of applicable group.
+                            if (member.isActive() && !isUserInException(publishedAssessment, member.getUserId(), site)) {
+                                log.debug("Calling send email notification: {}", member.getUserId());
+                                sendEmailNotification(site, publishedAssessment, member, null);
+                            }
+                        }
+                    }
+                } else if(StringUtils.equals(publishedAssessment.getAssessmentAccessControl().getReleaseTo(), "Anonymous Users")){
+                    //don't do anything when we've released to Anonymous Users.
+                } else {    //if there is no special ReleaseTo setting, we can email all site members.
+                    for (Member member : site.getMembers()) {
+                        log.debug("Samigo assessment open notification checking if member needs an email, member: {}", member.getUserId());
+                        log.debug("Member info: active? {}", member.isActive());
+                        log.debug("Member info: released to this member? {}", "TODO: THIS");
+                        if (member.isActive() && !isUserInException(publishedAssessment, member.getUserId(), site)) {   //send only to active members who aren't part of any exceptions
+                            // TODO: Add other checks here!
+                            log.debug("Calling send email notification: {}", member.getUserId());
+                            sendEmailNotification(site, publishedAssessment, member, null);
+                        }
+                    }
+                }
+            }
+        } catch (IdUnusedException e) {
+            log.warn("Samigo assessment available notification job failed, {}", e.getMessage(), e);
+        } finally {
+            sakaiSession.invalidate();
```

Comment:
This is where you would pop the advisor off the stack, rather than invalidating the session.